File: src/types/member.ts
// types/member.ts

/**
 * Statut d'un membre au sein de l'organisation.
 */
export enum MemberStatus {
  ACTIF = 'actif',
  INACTIF = 'inactif',
  SUSPENDU = 'suspendu'
}

/**
 * Filières d'études disponibles.
 */
export enum Filiere {
  GENIE_INFORMATIQUE = 'Génie Informatique',
  GENIE_CIVIL = 'Génie Civil',
  GENIE_ELECTRIQUE = 'Génie Électrique',
  GENIE_MECANIQUE = 'Génie Mécanique',
  MSP = 'MSP',
  AN = 'Arts Numérique',
  HN = 'Humanités Numériques',
  IFA = 'Finances et Actuariats',
  GENIE_TELECOM = 'Génie des Télécommunications'
}

/**
 * Niveaux d'études (Licence et Master).
 */
export enum Niveau {
  L1 = 'L1',
  MSP1 = 'MSP1',
  L2 = 'L2',
  MSP2 = 'MSP2',
  L3 = 'L3',
  M1 = 'M1',
  M2 = 'M2'
}

/**
 * Cellules (ou départements) internes de l'organisation.
 */
export enum Cellule {
  PROJET = 'Projet',
  COMMUNICATION = 'Communication',
  FORMATION = 'Formation',
  EVENEMENTIEL = 'Événementiel',
  FINANCE = 'Finance',
  COMPETITION = 'Compétition',
  RELATION_EXTERNE = 'Relation Externe'

}

/**
 * Interface représentant les données complètes d'un membre.
 */
export interface Member {
  id: string;
  numero_membre: string;
  nom: string;
  prenom: string;
  email: string;
  telephone: string;
  // Utilisation de l'énumération ou d'une chaîne pour la flexibilité (si l'API utilise des chaînes)
  filiere: Filiere | string;
  niveau: Niveau | string;
  cellule?: Cellule | string; // Optionnel
  date_adhesion: string; // Format ISO 8601 (e.g., "YYYY-MM-DD")
  date_expiration: string; // Format ISO 8601
  statut: MemberStatus | string;
  photo_url?: string; // Optionnel
  created_at: string; // Horodatage de création
  updated_at: string; // Horodatage de dernière mise à jour
}

/**
 * Interface représentant les métadonnées d'une carte de membre (physique ou numérique).
 */
export interface CardMetadata {
  id: string;
  member_id: string;
  recto_url: string;
  verso_url: string;
  qr_token: string;
  qr_expires_at: string; // Date d'expiration du token QR
  generated_at: string;
  download_count: number;
  last_downloaded_at?: string; // Optionnel
  verification_count: number;
  last_verified_at?: string; // Optionnel
  is_active: boolean;
}

/**
 * Interface représentant le résultat d'une vérification de carte (via QR code ou numéro).
 */
export interface VerificationResult {
  valid: boolean; // État global de la vérification
  member?: Member; // Données du membre si la vérification est réussie
  card?: { // Métadonnées de la carte pertinente
    generated_at: string;
    qr_expires_at: string;
  };
  verified_at: string; // Horodatage de la vérification
  reason?: string; // Raison de l'échec (si 'valid' est false)
  message?: string; // Message utilisateur/API supplémentaire
}

/**
 * Données requises pour l'inscription d'un nouveau membre.
 */
export interface MemberRegistration {
  nom: string;
  prenom: string;
  email: string;
  telephone: string;
  filiere: Filiere | string;
  niveau: Niveau | string;
  password: string;
}

/**
 * Données autorisées pour la mise à jour du profil par le membre lui-même.
 */
export interface MemberUpdate {
  nom?: string;
  prenom?: string;
  telephone?: string;
  photo_url?: string;
  filiere?: Filiere | string;
  cellule?: Cellule | string;
  niveau?: Niveau | string;
}

File: app/models/member.py
# backend/app/models/member.py

from pydantic import BaseModel, EmailStr, Field
from typing import Optional, Literal
from datetime import date, datetime
from enum import Enum

class MemberStatus(str, Enum):
    """Statuts possibles d'un membre"""
    ACTIF = "actif"
    INACTIF = "inactif"
    SUSPENDU = "suspendu"

class Filiere(str, Enum):
    """Filières disponibles"""
    GENIE_INFORMATIQUE = "Génie Informatique"
    GENIE_CIVIL = "Génie Civil"
    GENIE_ELECTRIQUE = "Génie Électrique"
    GENIE_MECANIQUE = "Génie Mécanique"

class Niveau(str, Enum):
    """Niveaux d'études"""
    L1 = "L1"
    L2 = "L2"
    L3 = "L3"
    M1 = "M1"
    M2 = "M2"

class Cellule(str, Enum):
    """Cellules du club"""
    PROJET = "Projet"
    COMMUNICATION = "Communication"
    FORMATION = "Formation"
    EVENEMENTIEL = "Événementiel"
    FINANCE = "Finance"

class MemberBase(BaseModel):
    """Données de base d'un membre"""
    numero_membre: str = Field(..., pattern=r"^CGI-\d{4}-\d{3}$")
    nom: str = Field(..., min_length=2, max_length=50)
    prenom: str = Field(..., min_length=2, max_length=50)
    email: EmailStr
    telephone: str = Field(..., pattern=r"^\+?\d{9,15}$")
    filiere: Filiere
    niveau: Niveau
    cellule: Optional[Cellule] = None
    date_adhesion: date
    statut: MemberStatus = MemberStatus.ACTIF

class MemberCreate(MemberBase):
    """Création d'un nouveau membre (Admin)"""
    photo_url: Optional[str] = None

class MemberRegistration(BaseModel):
    """Inscription de l'étudiant lui-même"""
    nom: str = Field(..., min_length=2, max_length=50)
    prenom: str = Field(..., min_length=2, max_length=50)
    email: EmailStr
    telephone: str = Field(..., pattern=r"^\+?\d{9,15}$")
    filiere: Filiere
    niveau: Niveau
    password: str = Field(..., min_length=8)

class MemberUpdate(BaseModel):
    """Mise à jour du profil par le membre"""
    nom: Optional[str] = Field(None, min_length=2, max_length=50)
    prenom: Optional[str] = Field(None, min_length=2, max_length=50)
    telephone: Optional[str] = Field(None, pattern=r"^\+?\d{9,15}$")
    filiere: Optional[Filiere] = None
    photo_url: Optional[str] = None
    cellule: Optional[Cellule] = None
    niveau: Optional[Niveau] = None

class MemberResponse(MemberBase):
    """Réponse avec données membre"""
    id: str
    photo_url: Optional[str] = None
    date_expiration: date
    created_at: datetime
    updated_at: datetime
    role: Literal["member", "admin"] = "member"
    
    class Config:
        from_attributes = True

MemberOut = MemberResponse 

class MemberCard(BaseModel):
    """Données pour la carte"""
    id: str
    numero_membre: str
    nom: str
    prenom: str
    filiere: str
    niveau: str
    cellule: Optional[str]
    date_adhesion: date
    date_expiration: date
    statut: str
    photo_url: Optional[str] = None
    qr_code_url: Optional[str] = None

class MemberInDB(MemberResponse):
    """Schéma complet d'un membre"""
    is_admin: bool = Field(False, description="Droits d'administration CARDIGI")
    permissions: list[str] = Field(default_factory=list, description="Liste des permissions détaillées")
    pending_sync: bool = Field(False, description="En attente de synchronisation avec l'API externe")

    class Config:
        from_attributes = True

class Token(BaseModel):
    """Schéma pour le JWT Access Token."""
    access_token: str
    token_type: str = "bearer"
    user: MemberInDB
    
class TokenData(BaseModel):
    """Schéma pour les données encodées dans le token."""
    sub: Optional[str] = None